src/Common/FitOnFhir.Common/Services/TokenValidationService.cs

```diff
@@ -0,0 +1,153 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.IdentityModel.Tokens.Jwt;
+using EnsureThat;
+using FitOnFhir.Common.Config;
+using FitOnFhir.Common.ExtensionMethods;
+using FitOnFhir.Common.Interfaces;
+using Microsoft.AspNetCore.Authentication.JwtBearer;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.IdentityModel.Protocols.OpenIdConnect;
+using Microsoft.IdentityModel.Tokens;
+
+namespace FitOnFhir.Common.Services
+{
+    public class TokenValidationService : ITokenValidationService
+    {
+        private IOpenIdConfigurationProvider _openIdConfigurationProvider;
+        private AuthenticationConfiguration _authenticationConfiguration;
+        private readonly IJwtSecurityTokenHandlerProvider _jwtSecurityTokenHandlerProvider;
+        private readonly ILogger _logger;
+        private Dictionary<string, string> _issuers;
+        private readonly List<Task> _configurationTasks = new List<Task>();
+
+        public TokenValidationService(
+            IOpenIdConfigurationProvider openIdConfigurationProvider,
+            AuthenticationConfiguration authenticationConfiguration,
+            IJwtSecurityTokenHandlerProvider jwtSecurityTokenHandlerProvider,
+            ILogger<TokenValidationService> logger)
+        {
+            _openIdConfigurationProvider = EnsureArg.IsNotNull(openIdConfigurationProvider, nameof(openIdConfigurationProvider));
+            _authenticationConfiguration = EnsureArg.IsNotNull(authenticationConfiguration, nameof(authenticationConfiguration));
+            _jwtSecurityTokenHandlerProvider = EnsureArg.IsNotNull(jwtSecurityTokenHandlerProvider, nameof(jwtSecurityTokenHandlerProvider));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _issuers = new Dictionary<string, string>();
```

Comment:
nit: The code might look cleaner if you moved this dictionary creation to the list of fields.

```c#
private readonly Dictionary<string, string> _issuers = new Dictionary<string, string>();
```